src/main/java/org/javamodularity/moduleplugin/tasks/PackageScanner.java

```diff
@@ -0,0 +1,65 @@
+package org.javamodularity.moduleplugin.tasks;
+
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Utility class for retrieving and collecting package names from source files (Java, Kotlin, Groovy, or Scala).
+ */
+public class PackageScanner {
+    private static final Logger LOGGER = Logging.getLogger(PackageScanner.class);
+
+    private static final String WS = "[ \t\r\n\\u000C]";
+    private static final String LETTER = "[a-zA-Z$_]|[^\\u0000-\\u007F\\uD800-\\uDBFF]|[\\uD800-\\uDBFF]|[\\uDC00-\\uDFFF]";
+    private static final String LETTER_OR_DIGIT = LETTER + "|[0-9]";
+    private static final String IDENTIFIER = "((" + LETTER + ")(" + LETTER_OR_DIGIT + ")*)";
+    private static final String QUALIFIED_NAME = IDENTIFIER + "(\\." + IDENTIFIER + ")*";
+
+    private static final String LINE_COMMENT = "//[^\r\n]*";
+    private static final String MULTILINE_COMMENT = "/\\*.*?\\*/";
+    private static final String IGNORE =  "(" + WS + "|" + LINE_COMMENT + "|" + MULTILINE_COMMENT + ")*";
+
+    private static final String PACKAGE_DECLARATION = "(?s)" + IGNORE + "package" + IGNORE + "(?<PACKAGE>" + QUALIFIED_NAME + ").*?";
+
+    private static final Pattern PATTERN = Pattern.compile(PACKAGE_DECLARATION);
+
+    private final Set<String> packages = new TreeSet<>();
+
+    public Set<String> getPackages() {
+        return packages;
+    }
+
+    /**
+     * Determines the package name of the specified source file and adds it to the {@link #packages} set.
+     * @param file the file to be scanned
+     * @return the package name, or null if the package declaration is missing or an error occurred.
+     */
+    public String scan(File file) {
+        if(!file.isFile()) {
+            LOGGER.warn("Not a source file: " + file);
+        } else {
+            try {
+                String text = new String(Files.readAllBytes(file.toPath()));
+                Matcher matcher = PATTERN.matcher(text);
+                if(matcher.matches()) {
+                    String packageName = matcher.group("PACKAGE");
+                    packages.add(packageName);
+                    return packageName;
+                } else {
+                    LOGGER.warn("Package declaration not found in: " + file);
+                }
+            } catch (Exception e) {
+                LOGGER.warn("Cannot scan " + file);
+                LOGGER.debug("Scan error", e);
+            }
```

Comment:
I'd follow the "fail fast" approach here. If we cannot parse a file and retrieve the package name, this means there will likely be a runtime error later on. Is there any reason to catch any exceptions occurred?

Reply:
I chose to issue warnings instead of failing, because the PackageScanner is not infallible and it also may deal with files that don't contain source code.

### 1. it is not infallible

While I expect the regex to correctly get the package names in almost all real-world scenarios, there are situations where it may fail to parse valid code:

**- The PackageScanner doesn't have a unicode preprocessor**

It cannot parse weird but valid package declarations such as this one:

```
// A valid package declaration: \u000d package org.copperengine.demo.jpms;
```
(JavaParser added such a [unicode preprocessor](https://github.com/javaparser/javaparser/pull/1928) only two weeks ago. By the author's own description, this "new provider is a horror".)


**- Kotlin allows [nested comments](https://kotlinlang.org/docs/reference/basic-syntax.html#comments)**

This is a valid comment in Kotlin:
```
/*
  Some comment
  /* Some nested comment */
*/
```
The PackageScanner cannot handle package declarations that contain nested comments, because Java's regex flavor doesn't support recursion or balancing groups.

**- Currently, the PackageScanner reads the whole content of the source file in a String.** 
This may be problematic with huge files.


### 2. it may deal with files that don't contain source code.
The PackageScanner currently scans all files in `testSourceSet.getAllSource()`, but not all files in this set contain proper source code. For example, the test-project has a [module-info.test](https://github.com/java9-modularity/gradle-modules-plugin/blob/master/test-project/greeter.provider/src/test/java/module-info.test) file in 'src/test/java'. The PackageScanner cannot find a package declaration in this file.

Reply:
Thank you for the detailed explanation. This absolutely makes sense.